k2/csrc/multisplit.h

```diff
@@ -0,0 +1,191 @@
+    uint32_t my_val = elements[i];
+    uint32_t my_bucket = bucket_mapping_func(my_val);
+    mark[i] = my_bucket;
+  }
+}
+
+/* MultiSplit using RadixSort.
+ *
+ * @param [in] context      Context indicating where the input data pointers,
+ *                          e.g., `in_keys` and `in_values`, come from.
+ * @param [in] num_elements Number of elements in `in_keys`, `in_values`,
+ *                          `out_keys` and `out_values`.
+ * @param [in] bucket_mapping_func
+ *                          It is a function like object support the following
+ *                          call operator: `uint32_t operator(uint32_t i)`.
+ *                          If it is NULL, we assume it is an identity function.
+ *                          That is, `in_keys[i]` also represents the bucket
+ *                          that `in_keys[i]` belongs to.
+ * @param [in] in_keys      Pointer to the input keys array.
+ * @param [in] in_values    Pointer to the input values array.
+ * @param [in] out_keys     Pointer to the output keys array. If not NULL,
+ *                          it will contain a sorted version of `in_keys`. The
+ *                          sort is performed according to the bucket
+ *                          information of the input elements and it is sorted
+ *                          in ascending order.
+ * @param [in] out_values   Pointer to the output values array. If not NULL,
+ *                          it will contain a sorted version of `in_values`.
+ */
+template <typename Lambda>
+void MultiSplitKeyValuePairs(ContextPtr context, int32_t num_elements,
+                             int32_t num_buckets, Lambda *bucket_mapping_func,
+                             const uint32_t *in_keys, const uint32_t *in_values,
+                             uint32_t *out_keys, uint32_t *out_values) {
+  K2_CHECK_GT(num_elements, 0);
+  K2_CHECK_GT(num_buckets, 0);
+  int32_t log_buckets = static_cast<int32_t>(ceilf(log2f(num_buckets)));
+  cudaStream_t stream = context->GetCudaStream();
+
+  constexpr int32_t kMultiSplitNumWarps = 8;
+  int32_t num_threads = kMultiSplitNumWarps * 32;
+  int32_t num_blocks = (num_elements + num_threads - 1) / num_threads;
+
+  Array1<uint32_t> d_mask(context, num_elements, 0);
+  const uint32_t *mask_data = in_keys;
+
+  if (bucket_mapping_func != nullptr) {
+    K2_CUDA_SAFE_CALL(MarkBinsGeneral<<<num_blocks, num_threads, 0, stream>>>(
+        in_keys, num_elements, *bucket_mapping_func, d_mask.Data()));
+    mask_data = d_mask.Data();
+  }
+
+  Array1<uint64_t> d_key_value_pairs(context, num_elements);
+  K2_CUDA_SAFE_CALL(
+      PackingKeyValuePairs<<<num_blocks, num_threads, 0, stream>>>(
+          num_elements, in_keys, in_values, d_key_value_pairs.Data()));
+
+  Array1<uint32_t> d_out(context, num_elements, 0);
+  size_t temp_storage_bytes = 0;
+
+  K2_CUDA_SAFE_CALL(cub::DeviceRadixSort::SortPairs(
+      nullptr, temp_storage_bytes, mask_data, d_out.Data(),
```

Comment:
Again, if `bucket_mapping_func == nullptr` , we can just calll `SortPairs`, there's no need to pack and unpack.

Reply:
But, for the case `bucket_mapping_func != nullptr`, I doubt if the caller really need `keys_out`, they may only cares the `values_out`. Anyway, they can always pass `mapped_keys` as keys and `Range(0, num_elems)` as values here, then they can gather `keys_out` and `values_out` with the re-ordered `Range(0, num_elems)` within only one lambda, much simpler than current `pack` and `unpack` approach.

Reply:
what if one of `out_keys` and `out_values`  is a nullptr?

Reply:
It's meaningless if a user sort key-value pairs but don't need the output, in that case (one of them is nullptr), what they should call is SortKeys